src/delta.c

```diff
@@ -38,86 +38,159 @@
+/* Check if the delta filename is well-formed, if so return true and fill the from/to
+ * buffers with the corresponding hashes. Return false otherwise. */
+static bool check_delta_filename(const char *delta_name, char *from, char *to)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
-	int ret;
-	struct stat sb;
+	/* Delta files have the form [FROM_VERSION]-[TO_VERSION]-[FROM_HASH]-[TO_HASH]. */
+	const char *s = delta_name;
+
+	/* Ignore versions, deltas will be used based on their hashes only. */
+	for (int i = 0; i < 2; i++) {
+		s = strchr(s, '-');
+		if (!s) {
+			return false;
+		}
+		s++;
+	}
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	/* Note: SWUPD_HASH_LEN accounts for the NUL-terminator after the hash. */
+	const size_t hash_len = SWUPD_HASH_LEN - 1;
+	if (strlen(s) != (hash_len * 2 + 1)) {
+		return false;
+	}
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	if (s[hash_len] != '-') {
+		return false;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
+	memcpy(from, s, hash_len);
 
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	/* Consume the first hash and the separator. */
+	s += hash_len + 1;
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
+	memcpy(to, s, hash_len);
 
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
-	xattrs_copy(origin, filename);
+	return true;
+}
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		char *to_staged = NULL;
+
+		char *delta_name = ent->d_name;
+		char *delta_file;
+		string_or_die(&delta_file, "%s/%s", delta_dir, delta_name);
+
+		char from[SWUPD_HASH_LEN] = {};
+		char to[SWUPD_HASH_LEN] = {};
+		if (!check_delta_filename(delta_name, &from[0], &to[0])) {
+			fprintf(stderr, "Invalid name for delta file: %s\n", delta_file);
+			goto next;
+		}
+
+		string_or_die(&to_staged, "%s/staged/%s", state_dir, to);
+
+		/* If 'to' file already exists, no need to apply delta. */
+		struct stat stat;
+		if (lstat(to_staged, &stat) == 0) {
+			goto next;
+		}
+
+		/* TODO: Sort the list of files by hash then walk it in parallel with readdir. */
+
+		struct list *ll = list_head(current_manifest->files);
+
+		struct file *found = NULL;
+		while (ll) {
```

Comment:
This can be simplified, 15 lines -> 12

		for (;ll;ll=ll->next) {
			struct file *file = ll->data;

			if (!file->is_deleted && !file->is_ghosted && file->is_file && hash_equal(file->hash, from)) {
		                apply_one_delta(file->filename, to_staged, delta_file, to);
				break;
			}
		}

		if (!ll) {
			fprintf(stderr, "Couldn't use delta file %s: no 'from' file to apply was found\n", delta_file);
		}

This code also misses the opportunity if there are multiple old files which match the hash, some of which are removed, to use one of the non-removed ones.

Reply:
Changed to use the for loop, but kept the found variable. We'd have to cast it anyway, potentially eating another line.

> This code also misses the opportunity if there are multiple old files which match the hash, some of which are removed, to use one of the non-removed ones.

> This code also misses the opportunity if there are multiple old files which match the hash, some of which are removed, to use one of the non-removed ones.

In the current code there's no difference between which one we pick, since we are processing these deltas before the removal happens. What kind of opportunity are you thinking about?

Reply:
I am thinking of something like

/bin/mail and /bin/rmail are the same file, but someone has moved /bin/mail to /bin/mail.old

Reply:
Got it.

Mixing this and the part of the other comment: added a check of the real filename hash before consider a file "found", this covers both the real file being delete or modified before the delta application starts.